### PR TITLE
File store freeze

### LINKED ISSUE #20288: Ensure file store mostly passes arq testsuite
This task is to change file / map store implementation so that it passes arquillian testsuite with the following exceptions:
- Ignore failed tests due to inability to change primary key (eg. client-id, username)
- Ignore failed tests due to not-yet supported concurrency (#15837)

### PR DESCRIPTION
**Pull request by hmlnarik**
This PR consists of several commits:
- [Fix ID determination](https://github.com/keycloak/keycloak/commit/0471b248ce085cc04c75679c3af9c23cb75314bb) adds a type variable for CRUD operations and consequently cleans up file structure @mhajas
- [Forbid changing ID (other setters)](https://github.com/keycloak/keycloak/commit/962d967463079acc4b322080914fff6bdae2cb41) fixes forgotten methods for writing potentially changing ID which were bypassed originally
- [Improve handling of null values](https://github.com/keycloak/keycloak/commit/a25b8d146858f8c236fe81497adf292e9411d255) correctly escapes `"null"` String in yaml files so that it is read as `"null"` and not `null`
- [Support convertible keys in maps](https://github.com/keycloak/keycloak/commit/bfbbaa478585b999efc20382db2777510d1a15a6) ensures that if map key is a non-string (e.g. `12` number), it is still read as `String` key
- [Fix writing empty values](https://github.com/keycloak/keycloak/commit/08d0b159eb28abab151e1d5dd94625871d84237c) ensures that empty value is written so that `""` is not read as `null`
- [Fix updated flag](https://github.com/keycloak/keycloak/commit/61d57eaf6ce1651f20ed097c8d74f18632f3c850) fixes delegates which could accidentally have swallowed `updated` flag instead of passing it on to delegated entity @mhajas
- [Proceed if an object has been deleted in the same tx](https://github.com/keycloak/keycloak/commit/d7f50b464a5ff1f004ed165712c58015f5ef1285) ensures that a removed-and-then-created file is not removed upon commit

Closes: #20288

### CODE REVIEW

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileCrudOperations.java [hunk 1]
```diff
@@ -210,6 +211,23 @@ public V create(V value) {
...
+
+        if (proposedId == null || proposedId.length == 0) {
+            return lastIdComponentIfUnset;
+        } else if (proposedId[proposedId.length - 1] == null) {
+            proposedId[proposedId.length - 1] = lastIdComponentIfUnset;
+        }
+
+        String[] escapedProposedId = escapeId(proposedId);
+        final String res = String.join(ID_COMPONENT_SEPARATOR, escapedProposedId);
+        if (LOG.isDebugEnabled()) {
```
Comment (mhajas) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileCrudOperations.java:225]:
```suggestion
        if (LOG.isTraceEnabled()) {
```

**Review by mhajas — COMMENTED** [review]
@hmlnarik What was the issue with the previous id determination? I see it is fixed but can't review it from the perspective of the wrong behaviour. 

Also, I added one comment to the code.

**Review by sguilhen — APPROVED** [review]

**Comment by hmlnarik** [comment]
@mhajas 

> What was the issue with the previous id determination? 

Object IDs are not stored inside of the YAML file, they are mostly determined from the object properties, e.g. client-id in clients. Now some IDs cannot be fully determined from non-id properties of the objects, e.g. [the permission ticket](https://github.com/hmlnarik/keycloak/blob/a2613151c11d6a48b8ff91d981e9a157c569851e/model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapStorageProviderFactory.java#L81). Such an ID can only then be constructed in the file store using the file name since the file only contains the object contents and not `id`. This was missing in the previous version.

This change makes it possible to check that a ID is preserved even for such objects which do store neither ID nor properties needed to construct the ID such as permission ticket in the object contents.

**Comment by hmlnarik** [comment]
Neither `WebAuthnSigningInTest` nor `InternationalizationTest` (#20348) failures are caused by this PR.

**Review by mhajas — APPROVED** [review]
Thank you @hmlnarik!

**Pull request merged**